File: py/edge.py
"""Edge technique — posterize-then-outline. See spec/edge.md.
Mirrors rs/src/edge.rs exactly. Reuses dilate_chebyshev from bands.py.
"""

from __future__ import annotations
import argparse
import math
import numpy as np
from PIL import Image
from bands import dilate_chebyshev


def aggr_to_n(aggr: float) -> int:
    """Map aggressiveness [0,100] -> posterize level count N [3,32] (round-half-up)."""
    return int(math.floor(3.0 + (aggr / 100.0) * 29.0 + 0.5))


def posterize_buckets(rgb_f32: np.ndarray, n: int) -> np.ndarray:
    """(H, W, 3) float32 -> (H, W, 3) int32 bucket indices [0..n-1], round-half-up.
    Promotes to float64 before multiply to match Rust's `c as f64 * (n-1)` path.
    """
    rgb64   = rgb_f32.astype(np.float64)
    buckets = np.floor(rgb64 * (n - 1) + 0.5).astype(np.int32)
    return np.clip(buckets, 0, n - 1)


def edge(rgba: np.ndarray, aggr: float, line_width: int) -> np.ndarray:
    """Compute edge layer for source rgba (H,W,4) float32.

    Returns a new (H,W,4) float32 array:
    edge pixels = (0,0,0,1) opaque black; others = (0,0,0,0) transparent.
    Source array is not modified.
    """
    h, w = rgba.shape[:2]
    n = aggr_to_n(aggr)

    # Step 1 — posterize
    bucks = posterize_buckets(rgba[..., :3], n)   # (H, W, 3) int32

    # Step 2 — 4-neighbour boundary detection (no wrap; border = same)
    # Pixel (y,x) is edge if any 4-neighbour has a different triple.
    diff_vert  = np.any(bucks[:-1, :] != bucks[1:,  :], axis=2)  # (H-1, W)
    diff_horiz = np.any(bucks[:, :-1] != bucks[:, 1:],  axis=2)  # (H, W-1)

    mask = np.zeros((h, w), dtype=bool)
    mask[:-1, :] |= diff_vert    # row y   differs from row y+1
    mask[1:,  :] |= diff_vert    # row y+1 differs from row y
    mask[:, :-1] |= diff_horiz   # col x   differs from col x+1
    mask[:, 1:]  |= diff_horiz   # col x+1 differs from col x

    # Step 3 — dilation (reuse Chebyshev from bands)
    r = line_width // 2
    if r > 0:
        mask = dilate_chebyshev(mask, r)

    # Step 4 — build output layer
    out = np.zeros((h, w, 4), dtype=np.float32)
    out[mask, 3] = 1.0   # alpha = 1; RGB stays 0 = opaque black

    return out


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("input")
    ap.add_argument("output")
    ap.add_argument("--aggr",  type=float, default=50.0)
    ap.add_argument("--width", type=int,   default=3)
    args = ap.parse_args()

    img  = Image.open(args.input).convert("RGBA")
    rgba = np.array(img, dtype=np.float32) / 255.0
    out  = edge(rgba, args.aggr, args.width)
    result = (out * 255.0 + 0.5).clip(0, 255).astype(np.uint8)
    Image.fromarray(result, "RGBA").save(args.output)
    print(f"Wrote {args.output}  N={aggr_to_n(args.aggr)}  r={args.width // 2}")


if __name__ == "__main__":
    main()

File: rs/src/edge.rs
//! Edge technique — posterize-then-outline. See spec/edge.md.
//! Reuses dilate_chebyshev from bands::dilate_chebyshev.

use crate::bands::dilate_chebyshev;

/// Map aggressiveness [0,100] to posterize level count N [3,32] (round-half-up).
pub fn aggr_to_n(aggr: f32) -> i32 {
    (3.0_f32 + (aggr / 100.0) * 29.0 + 0.5).floor() as i32
}

/// Posterize one f32 channel value [0,1] to bucket index 0..N-1 (round-half-up).
#[inline]
fn bucket(c: f32, n: i32) -> i32 {
    let x = c as f64 * (n - 1) as f64 + 0.5;
    (x.floor() as i32).clamp(0, n - 1)
}

/// Compute edge mask and return a new RGBA f32 layer buffer (W×H×4, row-major).
/// Edge pixels = opaque black (0,0,0,1). All others = transparent (0,0,0,0).
/// The source buffer is not modified.
///
/// `rgba` — source layer, f32 linear, W×H×4 row-major.
/// `aggr` — aggressiveness 0..100 (maps to N via `aggr_to_n`).
/// `line_width` — 1..7; dilation radius = line_width / 2.
pub fn edge(rgba: &[f32], width: usize, height: usize, aggr: f32, line_width: usize) -> Vec<f32> {
    let n    = aggr_to_n(aggr);
    let npix = width * height;

    // Step 1: posterize each pixel to (br, bg, bb) bucket triple
    let mut bucks: Vec<[i32; 3]> = Vec::with_capacity(npix);
    for i in 0..npix {
        let base = i * 4;
        bucks.push([
            bucket(rgba[base    ], n),
            bucket(rgba[base + 1], n),
            bucket(rgba[base + 2], n),
        ]);
    }

    // Step 2: 4-neighbour boundary detection (no wrap; border = same)
    let mut mask = vec![false; npix];
    for y in 0..height {
        for x in 0..width {
            let me = &bucks[y * width + x];
            let edge =
                (y > 0           && bucks[(y-1)*width + x    ] != *me) ||
                (y + 1 < height  && bucks[(y+1)*width + x    ] != *me) ||
                (x > 0           && bucks[y*width     + x - 1] != *me) ||
                (x + 1 < width   && bucks[y*width     + x + 1] != *me);
            if edge { mask[y * width + x] = true; }
        }
    }

    // Step 3: dilation (reuse Chebyshev from bands)
    let r = line_width / 2;
    let dilated = if r > 0 {
        dilate_chebyshev(&mask, width, height, r)
    } else {
        mask
    };

    // Step 4: build output layer (opaque black on edge, transparent elsewhere)
    let mut out = vec![0.0f32; npix * 4];
    for i in 0..npix {
        if dilated[i] {
            // out[i*4 + 0..2] stay 0.0 (black)
            out[i * 4 + 3] = 1.0; // opaque
        }
    }
    out
}
